### ISSUE TITLE
0.55.0: Static Over Final Class warning occurs on overridable class property

### ISSUE THREAD

**Issue by amcinnis** [realm/SwiftLint#5570]
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

In Xcode's boilerplate UITests launch tests code, they override a XCTest class property. This triggers a static over final class violation warning in SwiftLint 0.55.0.

```
final class StaticOverFinalUITestsLaunchTests: XCTestCase {

    // This triggers a static_over_final_class violation:
    override class var runsForEachTargetApplicationUIConfiguration: Bool {
        true
    }

   ...
   ..
   .
}
```

The [Static Over Final Class](https://realm.github.io/SwiftLint/static_over_final_class.html) documentation mentions the warning should only populate for non-overridable declarations. Is this a bug?

**Comment by tonyarnold** [comment]
I'm seeing the same in my project, in other NSObject-derived classes.

**Comment by SimplyDanny** [comment]
The intention of the "non-overridable" in the documentation is like "elements that cannot be overridden further should be marked `static` instead of `final class`".

In your example, `static` instead of `class` should be valid. Or are you seeing any compilation issues after the change?

**Comment by amcinnis** [comment]
Ah, thanks for that clarification Danny. If I replace `override class var` with `override static var` then everything compiles and SwiftLint does not show a violation warning. I believe I have a better understanding now of what the rule tries to accomplish.

Hopefully this GitHub issue will help others better understand as well!

**Comment by SimplyDanny** [comment]
Perhaps it's better to remove the part "for non-overridable declarations" from the message entirely. Anyone willing to do that? 😄

**Comment by amcinnis** [comment]
If I did, I would explicitly add this case to the static over final documentation for both triggering and non-triggering examples since it calls out Xcode boilerplate code that's bound to tigger a violation upon project creation.

However, my job prevents me from contributing to open source software so hopefully someone else can take the lead here 😅

**Comment by tonyarnold** [comment]
Yeah, the wording of the warning here isn't clear. I took "non-overridable declarations" to mean methods that weren't overridden.

**Comment by tonyarnold** [comment]
"Prefer static methods over class methods on final classes"?

For interest: I'm honestly unclear what the benefit is to this rule — is it just for consistency? (It would seem not, given that it's only catching final classes). 

Is there any functional difference between a class method and a static method on a final class?

**Comment by ZevEisenberg** [comment]
TIL: `static` is an alias for `final class`: https://stackoverflow.com/a/29206635/255489

This could be part of the confusion: I didn't know `final class` functions were a thing in Swift, so seeing that in the error message was confusing. Maybe leave it out if it was like the declaration I mentioned?

**Comment by SimplyDanny** [comment]
> TIL: `static` is an alias for `final class`: https://stackoverflow.com/a/29206635/255489

This is one aspect of the rule. They are both equivalent, so the rule suggests to prefer the shorter one. In final classes, `final` is implicitly applied to all contained declarations, thus every `class` becomes `final class`.

> Is there any functional difference between a class method and a static method on a final class?

I don't think there is, at least not in Swift. Perhaps there are both `class` and `static` in Swift because of backwards compatibility with Objective-C. But this is not my competence.

Anyway, due to having two keywords, the prescribed meaning for `class` is "can be overridden" as opposed to `static` which cannot. However, together with `final`, nothing can actually be overridden, hence the suggestion to use `static` instead.

Incidentally, the Swift compiler complains about `open` being used in `final` classes or declarations likewise.

My suggestions are "Prefer `static` over `final class`" on `f` in
```swift
class C1 {
  final class func f() {}
}
``` 
and "Prefer `static` over `class` in a final class" on `f` in
```swift
final class C2 {
  class func f() {}
}
```

Does that make things clearer?

**Comment by futuretap** [comment]
```
class A {
  open class var shared: A { aInstance }
  private static let aInstance = A()
}

class B: A {
  override class var shared: B { bInstance }
  private static let bInstance = B()
}

```

This doesn't work with `static`: `Static declarations are implicitly 'final'; use 'public' instead of 'open'`.
When replacing `open` with `public`: `Static property overrides a 'final' class property`

**Comment by SimplyDanny** [comment]
@futuretap: Where exactly in your example would the rule currently trigger? I'm wondering, because there is no `final class` anywhere.

**Comment by futuretap** [comment]
@SimplyDanny in the `override class var shared` line.

**Comment by SimplyDanny** [comment]
> [@SimplyDanny](https://github.com/SimplyDanny) in the `override class var shared` line.

But there's no `final` ...

**Comment by futuretap** [comment]
> But there's no `final` ...

There is a `Static declarations are implicitly 'final'; use 'public' instead of 'open'` build error if I follow the advice and move from class to static vars.
Then when replacing `open` with `public`, there is a `Static property overrides a 'final' class property` build error. Even though there’s no explicit `final`, I assume it's due to the "implicitly final" behavior mentioned before.

**Comment by futuretap** [comment]
Apologies, my example should be:

```
class AClass {
	open class var shared: AClass { aInstance }
	private static let aInstance = AClass()
}

final class BClass: AClass {
	override class var shared: BClass { bInstance }
	private static let bInstance = BClass()
}
```
(`AClass` instead of `A` to fix the `type_name` violation, **final** class BClass)

When I remove the `final` in `Bclass`, the warning goes away. However, I don’t see a reason why the `final class` is a problem. It gives the compiler a hint that this class won’t be subclassed.

**Comment by SimplyDanny** [comment]
Now the rule asks you to rewrite the second class like
```swift
final class BClass: AClass {
  override static var shared: BClass { bInstance }
  private static let bInstance = BClass()
}
```
which should not cause a compilation error.

**Comment by futuretap** [comment]
That fails with `Cannot override static property` if I make the `shared` var in `AClass` static as well. If I leave it as a class var in `AClass`, then it works. However, it feels strange to override a class var with a static var.

**Comment by SimplyDanny** [comment]
> That fails with `Cannot override static property` if I make the `shared` var in `AClass` static as well. If I leave it as a class var in `AClass`, then it works. However, it feels strange to override a class var with a static var.

Got that. However, the fact that `final class == static` is the purpose of the rule. `final class C { class var ... }` feels strange as well as `class var` intents overridability.
